mgmt: mcumgr: lib: cmd: dev: Add device information handler

nordicjm opened the pull request:
PR for adding a device information mcumgr group. Functionality is similar to the linux `uname` command where a format string can be provided to return certain information including board name, processor type, architecture, kernel name/version, OS name. It can be extended by other code using the revamped mcumgr callback system to add additional output/processing or to replace the OS name.

    The device information handler can be used to retrieve information about
    the configuration of the configured device such as board name, board
    revision, firmware version and build date.

Example output, no parameters:
`Zephyr`
'a' parameter, with default Bluetooth enabled settings:
`Zephyr Zephyr zephyr-v3.0.0-2569-g7f0654f969f7 3.0.99 arm unknown bt510 Zephyr`

- [X] Base implementation
- [X] Build date/time support (with note about invalidating reproducible builds, disabled by default)
- [X] Tests (check right response is received, check if memory error is received if buffer size is too small, check overriding OS name)
- [X] Release notes update
- [X] Documentation

This requires https://github.com/zephyrproject-rtos/zephyr/pull/50578 as a base and has pulled those commits into this review (first 7 commits of this PR can be ignored)

de-nordic reviewed (changes requested):
I think that this can be addition to OS group, there is no need to create additional group for that.
Consider also making the command more complex, like for example BSD `uname`, where request could be filled with optional "format" string that would be assigned letters representing requested info, for example
```
format="bfa";
```
would give response:
```
b="<CONFIG_BOARD>"
f="0.1.2-something"
unsupported="a"
```

lairdjm commented:
@de-nordic Moved to OS and uname-style output, still some parts to implement. Thoughts on this? Not sure if the processor response is already available or if it would become an #ifdef mess with many types of processors and processor families. Is there a way to currently get the overall package that you know of e.g. "Zephyr" for if purely using zephyr or "nRF connect SDK" if using that, etc.?

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/Kconfig:
```diff
@@ -118,7 +118,15 @@ config OS_MGMT_ECHO_LENGTH
 
 endif
 
+
```
Fix

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:
```diff
@@ -1,10 +1,13 @@
  */
 
+#include <stdio.h>
+
```
Fix newline

de-nordic commented:
> @de-nordic Moved to OS and uname-style output, still some parts to implement. Thoughts on this? Not sure if the processor response is already available or if it would become an #ifdef mess with many types of processors and processor families. Is there a way to currently get the overall package that you know of e.g. "Zephyr" for if purely using zephyr or "nRF connect SDK" if using that, etc.?

I do not think that we have a way to get processor name for now.
During compilation we do not know processor name only the arch, maybe processor name could be pulled out of DTS somehow.
We can probably postpone the processor part and complete all other "letters".

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/Kconfig:
```diff
@@ -121,4 +121,11 @@ endif
 config OS_MGMT_MCUMGR_PARAMS
 	bool "MCUMGR Parameters retrieval command"
 
+config OS_MGMT_UNAME
```
Invent your own name for this. Uname is Unix wide command and we have Zephyr here. Make some "os info", "zephyr info" or something. We can use similar option letters as "uname" but command should be ours.

lairdjm replied:
Fixed.

lairdjm commented:
Added a custom function to allow additional options in application, and changed OS to be it's own (weak) function to allow overriding by vendors

de-nordic commented:
@nordicjm Should we bring this back to live?

nordicjm commented:
Would be a good feature to add, yes.

de-nordic commented:
> Would be a good feature to add, yes.

Can we re-open this and set target to 3.3?

nordicjm committed:
doc: mgmt: smp group 0: Add OS info details

Adds details on the OS information mcumgr command.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release: 3.3: Add os_mgmt info command

Adds details of a new os_mgmt command for getting OS/application
information.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
mgmt: mcumgr: lib: cmd: os: Add device information handler

The device information handler can be used to retrieve information about
the configuration of the configured device such as board name, board
revision, firmware version and build date.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
tests: mgmt: mcumgr: os_mgmt_info: Add tests

Adds tests for the os_mgmt info command.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
mgmt: mcumgr: transport: dummy: Fix issue with large packets

Fixes an issue with large packets being received, these packets are
chunked into 127-byte frames for the serial transport but this system
is not needed for the dummy transport as it has a fixed size buffer.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:
```diff
@@ -302,6 +318,283 @@ os_mgmt_mcumgr_params(struct mgmt_ctxt *ctxt)
+		rc = snprintf(&output[output_length], (sizeof(output) - output_length), "%s%s",
+			      (prior_output == true ? " " : ""), STRINGIFY(BUILD_VERSION));
+#else
+		/* TODO: Get git branch or tag? */
+		rc = snprintf(&output[output_length], (sizeof(output) - output_length),
+			      "%sunknown", (prior_output == true ? " " : ""));
```
Needs addressing

de-nordic commented on tests/subsys/mgmt/mcumgr/os_mgmt_info/src/smp_test_util.c:
```diff
@@ -0,0 +1,44 @@
+		.nh_id = OS_MGMT_ID_INFO,
+	};
+
+	mgmt_hton_hdr(rsp_hdr);
```
This is internal function of SMP(rotocol) processing that should have never been exported, but mgmt interface is a mess so it was. It is even named wrongly.
This test implements client so it can spare some additional line to implement the client side of header encoding.

nordicjm replied:
Fixed

nordicjm commented:
Rebased

de-nordic reviewed (changes requested):
One comment regarding Kconfig names.

nordicjm commented:
@de-nordic Changes made, left some of the text as-is to prevent duplicating strings

de-nordic reviewed (approved):
I do not like the `(prior_output == true ... ` things in `sprintf` as I think that saving on strings will not be so significant with branching overhead.
Having said that I do not find it a deal breaker, as this is something we can argue on later and improve (or not).

Looks OK, Thanks.

fabiobaltieri commented:
> @fabiobaltieri can you merge this please?

Needs a second review, I'll take a look.

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt/include/os_mgmt_processor.h:
```diff
@@ -0,0 +1,137 @@
+ * Will be unknown if processor type is not listed
+ * (List extracted from /cmake/gcc-m-cpu.cmake)
+ */
+#if defined(CONFIG_ARM)
```
Indeed it is horrible to look at and update, indentation would probably be a good idea, I'll give it a try and see if CI throws an error about it

fabiobaltieri replied:
Wondering if it's even worth reporting it in such details, maybe you could just report whatever is in `cpu@0` compatible.
